### ISSUE TITLE
Table class addition

### ISSUE THREAD

**Issue by jreadey** [h5py/h5py#2095]
I was wondering if the h5py community would be interested in Dataset derived class specifically for working with tabular data (by which I mean 1-dimensional, compound-type datasets).

I did this for h5pyd, you can see the code here: https://github.com/HDFGroup/h5pyd/blob/master/h5pyd/_hl/table.py.

It doesn't really add any new capability over the Dataset class, but is a bit more convenient  for working with 1D datasets.

Example:

    with h5py.File("myfile.h5") as f:
        table = f["mytable"]. # a Table class is returned if mytable is 1-d, compound
        print(f"mytable has {table.nrows} rows")
        cursor = table.create_cursor()  # The cursor will retrieve a buffer of rows as needed for efficiency 
        for row in cursor:
             print(row)

        query = "field_A == 42"
        rows = table.read_where(query). # returns numpy array of rows matching the condition
        print(f"got {len(rows)} where field_A is 42")

        more_rows = [(32, "apple", 9.3), (39, "banana", 3.2), (30, "orange", 6.5)]
        table.append(more_rows). # extends dataspace and copies data

Anyway, if there is interest, I'd be happy to submit a PR for a h5py-version of the h5pyd Table class.

**Comment by tacaswell** [comment]
I like the idea, but I am not stoked about the auto-magical detection.  

Could we do it as as accessor like

```python
table = f['mytable'].as_table()    # explodes if not 1-d and compound
```

?

Would it also make sense to put a past-path to pandas on this object (with a suitable buried import)?

How hard would it be to make the return value of `table.read_where` be the same type as returned by `table.create_cursor()`? Could we think of `create_cusor` as `read_where("True")`? 

If we go down the path of taking string expressions and evaluating them, how far down that path do we go? `"A > 3"`  `"A < B"` v "A/B < C"` v "log(A/B) > C" etc?

**Comment by aragilar** [comment]
Yeah, auto-magical stuff is good until it doesn't do what you want...

How would this relate to other projects which provide tabular access/interfaces, such as pytables, pandas or astropy tables? The example code doesn't look too complex, but would it create some level of duplication of effort?

**Comment by takluyver** [comment]
I'm less positive on this, because I think what you're proposing sounds very similar to [pytables](https://www.pytables.org/index.html) - see in particular the [Table.where()](www.pytables.org/usersguide/libref/structured_storage.html#tables.Table.where) method. Pytables has already gone much further in this direction, e.g. it can manage and use indexes for table columns. I can see the attraction of having a relatively 'simple' table API in h5py, but I don't think there's any obvious way to define what's in scope for that and what's not, so there would always be pressure to add features.

I think h5py is better off sticking to its niche of providing a Pythonic but relatively plain API to HDF5 storage, with very little extra functionality on top of that. I don't think there's anything to stop someone making this API as a separate package to use with h5py (starting with something like `table = H5Table(f['mytable'])`.

**Comment by ajelenak** [comment]
I agree there is a fine line but I am in favor in including something like this in h5py. (Disclaimer: I work for the same company as @jreadey.)

I don't know about AstroPy tables, but both PyTables and Pandas are for serious tabular data users and encode their data in a very specific way in HDF5 files. What is proposed here would apply only to  1D compound HDF5 datasets, in files that otherwise may contain all kinds of different types of data as well.

DImension Scale API is designated as high-level, thus not part of the core HDF5 API, but it is supported in h5py. I think this Table class is similar in purpose.

Fending off possible future requests for advanced features should not be difficult because of the aforementioned packages. We are here discussing a meaningful baseline.

**Comment by jreadey** [comment]
My experience as a h5py-user is that when I need to work with tabular data I either:
1) just do things in a slightly cumbersome way necessitated by h5py's added complexity needed to support multi-dimensional data
 2) add helper functions similar to what's in the proposed table class 
3) or.. if the dataset is small enough, read into a pandas dataframe and work from there

I wouldn't convert my file to pytables and start from there though.  I guess the hope was for a unified h5py+pytables, but like @tacaswell said, that never got completed.

I wouldn't expect the Table class to ever extend to the point where it needed ancillary data to be stored  in the file (like indexes), but the given functionality would seem convenient for h5py-ers working with tabular data - but maybe that's just me...  I'm happy to go with the community consensus.

**Comment by takluyver** [comment]
> Fending off possible future requests for advanced features should not be difficult because of the aforementioned packages. 

My experience is that that's not really how things go. When there's a big, powerful option and a small, simpler one, you hear from a constant succession of people who really want to use the simpler one, but it's just lacking this one little piece they need, and it would be so easy to add it. So you do, and they're happy, but then someone else would use it if only X was added, and so on.

> I wouldn't expect the Table class to ever extend to the point where it needed ancillary data to be stored in the file (like indexes),

Thanks, if it does go in, that sounds like a useful rule for limiting the scope.

> I'm happy to go with the community consensus.

So far there seem to be more people in favour than opposed, and I'll go with the consensus if it's against me. :slightly_smiling_face: 

One more plug for a separate package on top of h5py: I haven't looked into your code in h5pyd at all, but might it be possible to have a table package that can work with either h5py or h5pyd? That might avoid duplicating some code.

**Comment by aragilar** [comment]
I'm fairly neutral on this (I can see the utility, but I agree with Tom about scope creep), though here's two ideas to mull over.

I recall that https://portal.hdfgroup.org/display/HDF5/Tables and https://portal.hdfgroup.org/display/HDF5/Packet+Tables exist for working with tabular data, so maybe we should wrap those, and then have a high level API over that (not sure how widely those APIs are used, but FITS tables are quite common in astronomy)?

astropy provides a "registry" system (https://docs.astropy.org/en/latest/io/registry.html), whereby you can load data from many different formats into classes. I wonder if borrowing some of the concepts from it might make sense? For example, `f["mytable"].load_as("table")` would look up the loader for a dataset labelled as `"table"` (and you could do something similar for groups). This might be preferable to adding `as_table`, which could result in people asking for `as_pandas`, `as_pytables` etc. This would then allow separate packages to provide higher level interfaces without needing significant integration work on the h5py side.

**Comment by takluyver** [comment]
I'd also like to avoid making some sort of plugin system for `tbl = f['mydataset'].load_as('table')` to load and use an external package, because I think using wrapper classes explicitly (like `tbl = H5Table(f['mydataset'])`) is clearer and will work better with tools like IDEs - an IDE can easily know what class `tbl` is and use it for completions, code checks etc.

**Comment by tacaswell** [comment]
I agree with @takluyver that plugins get messy fast.

It is not obvious to me that you could write code that would be performant for both h5py and h5pyd because it will depend on the details and we may want to drop directly to the low-level API.

I also agree with @aragilar that having a version that works on groups rather than structured data would be interesting.

**Comment by ajelenak** [comment]
> I also agree with @aragilar that having a version that works on groups rather than structured data would be interesting.

+1. I've seen and used this approach to mimic columnar storage of tabular data in HDF5. I think the rules for this should be simple:

1. Any 1D non-compound HDF5 dataset in the group.
2. Of those, the dataset with the largest number of elements defines the number of "table's" rows.
3. Fill values are per dataset. And returned also in cases when the requested rows go beyond the dataset's size.

Or something better!

**Comment by tacaswell** [comment]
> 2. Of those, the dataset with the largest number of elements defines the number of "table's" rows.

I would have said the simple rule is either the user specifies the length (in which case columns are either filled or trimmed as needed) or raises if they are not all the same length! 

I am coming to:

 - this is a useful thing that should exist
 - should start life as an external package.  I would be happy to host it in the h5py org (but given the hdf5 groups involvement that org).  It would be good start with the theory it can be generic between h5py and h5pyd and prove that wrong (@jreadey might already know one way or the other).  starting stand alone means we can version rev it fast (It would also make it possible to pick up a pandas dependency!) and be a lot more fluid with back-compatibility (if it goes out in h5py we can not take it back).
 - I could be sold on it being built in to h5py as an accesor on dataset/group once we are sure it is right / if we need to reach into the guts to get the performance to make it worth it.

**Comment by takluyver** [comment]
That's a good example of the kind of complexity I'm concerned about: we've already proposed two kinds of tables which could be superficially similar in the user facing API, even though they're stored in very different ways. Each option has its strengths, so there isn't an obvious way to decide to support one and not the other. Then there's the potential for confusion when you can only do X with one kind of table, or you can do Y but it's much more efficient with one than the other...

**Comment by jreadey** [comment]
Based on the comments here, I'll start with a separate package for table support that works with h5py or h5pyd.  Once that's done, we'll see how it evolves and whether it would eventually make sense to integrate with h5py or not.
Thanks for everyone's feedback!

**Comment by ivirshup** [comment]
A little late to the party, but I would share how [`anndata` does tables in hdf5 and zarr](https://anndata.readthedocs.io/en/latest/fileformat-prose.html#dataframes). Basically it's a columnar format, and we crib from arrow. Examples of how we steal from arrow include:

* Categorical dtype (group with categories and codes)
* Nullable types (group with data and mask)
* Considering doing ragged like arrow's lists

These are fairly easy to load as dask dataframes ([gist](https://gist.github.com/ivirshup/884df705592186a55b56c2c863e1f884)), and [`anndata-rs`](https://github.com/kaizhang/anndata-rs) will load from these into polars dataframes, which I believe uses arrow memory representation.

One thing that would make this format better is variable length chunks so we could index on low cardinality features (similar to [parquet](https://arrow.apache.org/docs/python/parquet.html#partitioned-datasets-multiple-files)).
